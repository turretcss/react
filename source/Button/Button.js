import React from 'react'
import classnames from 'classnames'

import { indicatorMap, sizeAliases } from '../constants'

const Button = ({
  Tag = 'button',
  type = 'button',
  className,
  size,
  border,
  square,
  color,
  pill,
  spinner,
  block,
  children,
  text,
  ...props
}) => (
  <Tag
    {...props}
    type={type}
    className={
      classnames(className, {
        ['button-' + (sizeAliases[size] || size)]: size,
        ['button-' + color]: color && !indicatorMap[color],
        ['button-block']: block,
        ['button-border']: border,
        ['button-pill']: pill,
        ['button-square']: square,
        ['spinner']: spinner,
        [color]: indicatorMap[color],
      }) || null
    }
    children={children || text}
  />
)

export const ButtonLink = props => <Button {...props} Tag="a" className={['button', props.className]} type={null} />

export default Button
