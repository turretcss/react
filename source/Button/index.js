export { default } from './Button'
export { default as Button, ButtonLink } from './Button'
