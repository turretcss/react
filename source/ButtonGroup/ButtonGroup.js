import React from 'react'
import classnames from 'classnames'

const ButtonGroup = ({ className, block, size, pill, square, color, border, disabled, children = [], ...props }) => (
  <div {...props} className={classnames('button-group', className, { 'button-group-block': block })}>
    {React.Children.toArray(children).map(child =>
      React.cloneElement(child, {
        size: child.props.size === undefined ? size : child.props.size,
        pill: child.props.pill === undefined ? pill : child.props.pill,
        square: child.props.square === undefined ? square : child.props.square,
        color: child.props.color === undefined ? color : child.props.color,
        border: child.props.border === undefined ? border : child.props.border,
        disabled: child.props.disabled === undefined ? disabled : child.props.disabled,
      })
    )}
  </div>
)

export default ButtonGroup
