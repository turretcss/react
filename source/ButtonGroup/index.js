export { default } from './ButtonGroup'
export { default as ButtonGroup } from './ButtonGroup'
