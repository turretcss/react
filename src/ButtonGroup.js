import React from 'react'
import classnames from 'classnames'

const ButtonGroup = ({ className, block, size, pill, square, color, border, children = [], ...props }) => (
  <div {...props} className={classnames('button-group', className, { 'button-group-block': block })}>
    {React.Children.toArray(children).map(child =>
      React.cloneElement(child, {
        size: size || child.props.size,
        pill: pill || child.props.pill,
        square: square || child.props.square,
        color: color || child.props.color,
        border: border || child.props.border,
      })
    )}
  </div>
)

export default ButtonGroup
