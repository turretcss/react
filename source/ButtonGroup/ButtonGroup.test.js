import React from 'react'
import renderer from 'react-test-renderer'
import { renderToStaticMarkup } from 'react-dom/server'

import ButtonGroup from './ButtonGroup'
import Button from '../Button'

describe('ButtonGroup', () => {
  it('renders', () => {
    const result = renderToStaticMarkup(<ButtonGroup />)
    const expected = '<div class="button-group"></div>'
    expect(result).toEqual(expected)
  })

  it('renders block prop', () => {
    const result = renderToStaticMarkup(<ButtonGroup block />)
    const expected = '<div class="button-group button-group-block"></div>'
    expect(result).toEqual(expected)
  })

  it('renders className prop', () => {
    const result = renderToStaticMarkup(<ButtonGroup className="custom" />)
    const expected = '<div class="button-group custom"></div>'
    expect(result).toEqual(expected)
  })

  it('renders children', () => {
    const result = renderToStaticMarkup(
      <ButtonGroup>
        <Button />
      </ButtonGroup>
    )
    const expected = '<div class="button-group"><button type="button"></button></div>'
    expect(result).toEqual(expected)
  })

  it('handles false children', () => {
    const result = renderToStaticMarkup(<ButtonGroup>{false}</ButtonGroup>)
    const expected = '<div class="button-group"></div>'
    expect(result).toEqual(expected)
  })

  it('handles false children alongside non-false children', () => {
    const result = renderToStaticMarkup(
      <ButtonGroup>
        <Button />
        {false}
      </ButtonGroup>
    )
    const expected = '<div class="button-group"><button type="button"></button></div>'
    expect(result).toEqual(expected)
  })

  it('spreads unknown props to the div', () => {
    const result = renderToStaticMarkup(
      <ButtonGroup data-custom-attr="Yeh">
        <Button />
      </ButtonGroup>
    )
    const expected = '<div data-custom-attr="Yeh" class="button-group"><button type="button"></button></div>'
    expect(result).toEqual(expected)
  })

  it('assigns button props to the children', () => {
    const tree = renderer
      .create(
        <ButtonGroup color="primary" size="s" pill square border disabled>
          <Button />
          <Button />
          <Button />
        </ButtonGroup>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('child props take priority over parent props', () => {
    const tree = renderer
      .create(
        <ButtonGroup size="s" color="primary" disabled pill border square>
          <Button />
          <Button size="l" color="secondary" disabled={false} pill={false} border={false} square={false} />
          <Button />
        </ButtonGroup>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
