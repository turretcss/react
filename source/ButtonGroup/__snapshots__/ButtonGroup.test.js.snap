// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup assigns button props to the children 1`] = `
<div
  className="button-group"
>
  <button
    className="button-s button-primary button-border button-pill button-square"
    disabled={true}
    type="button"
  />
  <button
    className="button-s button-primary button-border button-pill button-square"
    disabled={true}
    type="button"
  />
  <button
    className="button-s button-primary button-border button-pill button-square"
    disabled={true}
    type="button"
  />
</div>
`;

exports[`ButtonGroup child props take priority over parent props 1`] = `
<div
  className="button-group"
>
  <button
    className="button-s button-primary"
    disabled={true}
    type="button"
  />
  <button
    className="button-l button-secondary"
    disabled={false}
    type="button"
  />
  <button
    className="button-s button-primary"
    disabled={true}
    type="button"
  />
</div>
`;
